src/nginxconfig/util/vue_backwards_compatibility.js

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2021 DigitalOcean
+
+This code is licensed under the MIT License.
+You may obtain a copy of the License at
+https://github.com/digitalocean/nginxconfig.io/blob/master/LICENSE or https://mit-license.org/
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions :
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+import isObject from './is_object';
+
+// Handle converting the old query format to our new query params
+export default data => {
+    // Handle converting old domain settings to new ones
+    if ('global' in data && isObject(data.global)) {
+        // Handle specifics global data
+        const mappedData = {
+            php: {},
+        };
+
+        // Keys to map
+        const keysToMap = {
+            php: [
+                'phpServer',
+                'phpServerCustom',
+                'phpBackupServer',
+                'phpBackupServerCustom',
+            ],
+        };
+
+        for (const key in data.global) {
+            if (!Object.prototype.hasOwnProperty.call(data.global, key)) continue;
+
+            // Handle all includes php key
+            if (key === 'php'){
+                for (const domainKey in data.domains) {
+                    if (!Object.prototype.hasOwnProperty.call(data.domains, domainKey)) continue;
+
+                    for (const valueKey in data.global[key]){
+                        if (!Object.prototype.hasOwnProperty.call(data.global[key], valueKey)) continue;
+
+                        if (keysToMap.php.includes(valueKey)) {
+                            mappedData[key][valueKey] = data.global[key][valueKey];
+                        }
+                    }
+
+                    // Overwrite mapped properties
+                    data.domains[domainKey] = {...data.domains[domainKey], ...mappedData};
+                }
```

Comment:
Why are we looping over domains here?

We should just be looping over `data.global` and `data.global[key]` to build out the `mappedData`.

Once those two nested loops have completed, you can then have a loop over `data.domains` to apply `mappedData`.

---

`data.domains[domainKey] = {...data.domains[domainKey], ...mappedData};` is going to cause issues if a domain already has some PHP settings, as those will be overwritten by the spread -- this needs to be a deep merge of the objects.

Reply:
Really this overwrite others options. We can create a recursive deepMerge function:
```javascript
import isObject from './is_object';

// Handle converting the old query format to our new query params
export default data => {
    // Handle converting old domain settings to new ones
    if ('global' in data && isObject(data.global)) {
        // Handle specifics global data
        const mappedData = {
            php: {},
        };

        // Keys to map
        const keysToMap = {
            php: [
                'phpServer',
                'phpServerCustom',
                'phpBackupServer',
                'phpBackupServerCustom',
            ],
        };

        for (const key in data.global) {
            if (!Object.prototype.hasOwnProperty.call(data.global, key)) continue;

            for (const key2 in data.global[key]) {
                if (!Object.prototype.hasOwnProperty.call(data.global[key], key2)) continue;

                if (keysToMap.php.includes(key2)) {
                    mappedData[key][key2] = data.global[key][key2];
                }
            }
        }

        // Deep fusion merge
        for (const key in data.domains) {
           if (!Object.prototype.hasOwnProperty.call(data.domains, key)) continue;
            deepMerge(data.domains[key], mappedData)
        }
    }
};

function deepMerge(target, source) {
    Object.keys(source).forEach(function (key) {
        if (source[key] && typeof source[key] === 'object') {
            merge(target[key] = target[key] || {}, source[key]);
            return;
        }
        target[key] = source[key];
    });
}
```
Approved ?

Reply:
**Really !** 
This overwrite others options. We can create a recursive deepMerge function:
```javascript
import isObject from './is_object';

// Handle converting the old query format to our new query params
export default data => {
    // Handle converting old domain settings to new ones
    if ('global' in data && isObject(data.global)) {
        // Handle specifics global data
        const mappedData = {
            php: {},
        };

        // Keys to map
        const keysToMap = {
            php: [
                'phpServer',
                'phpServerCustom',
                'phpBackupServer',
                'phpBackupServerCustom',
            ],
        };

        for (const key in data.global) {
            if (!Object.prototype.hasOwnProperty.call(data.global, key)) continue;

            for (const key2 in data.global[key]) {
                if (!Object.prototype.hasOwnProperty.call(data.global[key], key2)) continue;

                if (keysToMap.php.includes(key2)) {
                    mappedData[key][key2] = data.global[key][key2];
                }
            }
        }

        // Deep merge
        for (const key in data.domains) {
           if (!Object.prototype.hasOwnProperty.call(data.domains, key)) continue;
            deepMerge(data.domains[key], mappedData)
        }
    }
};

function deepMerge(target, source) {
    Object.keys(source).forEach(function (key) {
        if (source[key] && typeof source[key] === 'object') {
            deepMerge(target[key] = target[key] || {}, source[key]);
            return;
        }
        target[key] = source[key];
    });
}
```
Approved ?

Reply:
**Really !** 
This overwrite others options. We can create a recursive deepMerge function:
```javascript
import isObject from './is_object';

// Handle converting the old query format to our new query params
export default data => {
    // Handle converting old domain settings to new ones
    if ('global' in data && isObject(data.global)) {
        // Handle specifics global data
        const mappedData = {
            php: {},
        };

        // Keys to map
        const keysToMap = {
            php: [
                'phpServer',
                'phpServerCustom',
                'phpBackupServer',
                'phpBackupServerCustom',
            ],
        };

        for (const key in data.global) {
            if (!Object.prototype.hasOwnProperty.call(data.global, key)) continue;

            for (const key2 in data.global[key]) {
                if (!Object.prototype.hasOwnProperty.call(data.global[key], key2)) continue;

                if (keysToMap.php.includes(key2)) {
                    mappedData[key][key2] = data.global[key][key2];
                }
            }
        }

        // Deep merge
       const deepMerge = function(target, source) {
          Object.keys(source).forEach(function (key) {
              if (source[key] && typeof source[key] === 'object') {
                    deepMerge(target[key] = target[key] || {}, source[key]);
                    return;
                }
                  target[key] = source[key];
              });
          }

        for (const key in data.domains) {
           if (!Object.prototype.hasOwnProperty.call(data.domains, key)) continue;

           deepMerge(data.domains[key], mappedData)
        }
    }
};
```
Approved ?